File: src/main/java/ru/ioleksiv/telegram/bot/core/api/model/method/message/ForwardMessage.java
package ru.ioleksiv.telegram.bot.core.api.model.method.message;

import com.fasterxml.jackson.annotation.JsonProperty;
import ru.ioleksiv.telegram.bot.core.api.model.objects.std.Message;
import ru.ioleksiv.telegram.bot.core.api.model.NetworkError;
import ru.ioleksiv.telegram.bot.core.model.method.ChatAction;
import ru.ioleksiv.telegram.bot.core.model.responses.CommonResponse;
import ru.ioleksiv.telegram.bot.core.model.responses.ResponseCollection;
import ru.ioleksiv.telegram.bot.core.controller.network.Networker;

/**
 * @see <a href="https://core.telegram.org/bots/api#forwardmessage">forwardMessage</a>
 */
public class ForwardMessage extends ChatAction<Message> {
    private static final String FORWARD_MESSAGE = "forwardMessage";
    /**
     * from_chat_id	Integer or String Unique identifier for the chat where the original message
     * was sent (or channel username in the format @channelusername)
     */
    @JsonProperty("from_chat_id")
    private Long fromChatId = null;
    /**
     * disable_notification	Boolean	Optional	Sends the message silently. Users will receive a
     * notification with no sound.
     */
    @JsonProperty("disable_notification")
    private Boolean disableNotification = null;
    /**
     * message_id	Integer Message identifier in the chat specified in from_chat_id
     */
    @JsonProperty("message_id")
    private Long messageId = null;

    public ForwardMessage(Networker networker) {
        super(FORWARD_MESSAGE, networker);
    }

    public Long getFromChatId() {
        return fromChatId;
    }

    public ForwardMessage setFromChatId(Long fromChatId) {
        this.fromChatId = fromChatId;
        return this;
    }

    public Boolean getDisableNotification() {
        return disableNotification;
    }

    public ForwardMessage setDisableNotification(Boolean disableNotification) {
        this.disableNotification = disableNotification;
        return this;
    }

    public Long getMessageId() {
        return messageId;
    }

    public ForwardMessage setMessageId(Long messageId) {
        this.messageId = messageId;
        return this;
    }

    @Override
    protected Class<? extends CommonResponse<Message>> getResultWrapperClass() {
        return ResponseCollection.MessageResponse.class;
    }

    @Override
    public ForwardMessage setChatId(Long chatId) {
        pSetChatId(chatId);
        return this;
    }

    @Override
    public ForwardMessage setNetworkErrorListener(NetworkError onNetworkError) {
        pSetNetworkErrorListener(onNetworkError);
        return this;
    }
}

File: server-api/src/main/kotlin/io/github/oleksivio/tl/kbot/server/api/method/message/ForwardMessage.kt
package io.github.oleksivio.tl.kbot.server.api.method.message

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonProperty
import io.github.oleksivio.tl.kbot.server.api.model.ApiDict
import io.github.oleksivio.tl.kbot.server.api.model.MessageResponse
import io.github.oleksivio.tl.kbot.server.api.model.method.common.ChatAction
import io.github.oleksivio.tl.kbot.server.api.objects.ChatId
import io.github.oleksivio.tl.kbot.server.api.objects.std.Message

/**
 * @see [forwardMessage](https://core.telegram.org/bots/api/#forwardmessage)
 */
data class ForwardMessage(
    /**
     * chat_id Integer or String Yes Unique identifier
     */
    @JsonProperty(ApiDict.CHAT_ID_KEY)
    override val chatId: ChatId,
    /**
     * from_chat_id Integer or String Yes Unique identifier for the chat where the original message was sent
     */
    @JsonProperty(ApiDict.FROM_CHAT_ID_KEY)
    val fromChatId: Long,
    /**
     * disable_notification [Boolean] OptionalSends the message silently.
     * Users will receive a notification with no sound.
     */
    @JsonProperty(ApiDict.DISABLE_NOTIFICATION_KEY)
    val disableNotification: Boolean? = null,
    /**
     * message_id Integer Yes Message identifier in the chat specified in from_chat_id
     */
    @JsonProperty(ApiDict.MESSAGE_ID_KEY)
    val messageId: Long

) : ChatAction<Message>() {

    @JsonProperty(ApiDict.METHOD_KEY)
    override val method = "forwardMessage"
    @JsonIgnore
    override val resultWrapperClass = MessageResponse::class
}
